types/validator_set_test.go

```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+	v3 = newValidator([]byte("v3"), 3)
+	v5 := newValidator([]byte("v4"), 55)
+
+	chVals = []*Validator{v3, v5}
+	assert.Nil(t, vals1.UpdateWithChangeSet(chVals))
+	assert.Nil(t, verifyValidatorSet(vals1))
+
+	chVals = []*Validator{v5, v3}
+	assert.Nil(t, vals2.UpdateWithChangeSet(chVals))
+	assert.Nil(t, verifyValidatorSet(vals2))
+
+	assert.Nil(t, verifyValidatorSetEquiv(vals1, vals2))
+
+}
+
+func verifyValidatorEquiv(val1, val2 *Validator) error {
+	if !bytes.Equal(val1.Address, val2.Address) || val1.VotingPower != val2.VotingPower {
+		return fmt.Errorf("validators %s and %s are different", val1, val2)
+	}
+	return nil
+}
+
+func verifyValidatorSetEquiv(vals1, vals2 *ValidatorSet) error {
+	if len(vals1.Validators) != len(vals2.Validators) {
+		return fmt.Errorf("validator sets %s, %s don't have same num of vals", vals1, vals2)
+	}
+	if getTotalVotingPower(vals1) != getTotalVotingPower(vals2) {
+		return fmt.Errorf("validator sets %s, %s don't have the same total voting power", vals1, vals2)
+	}
+	if getTotalProposerPriority(vals1) != getTotalProposerPriority(vals2) {
+		return fmt.Errorf("validator sets %s, %s don't have the same total priorities", vals1, vals2)
+	}
+	for i, val1 := range vals1.Validators {
+		val2 :=  vals2.Validators[i]
+		if err := verifyValidatorEquiv(val1, val2); err != nil {
+			return fmt.Errorf("validators at index %d are different: %s", i, err)
+		}
+	}
+	return nil
+}
+
+func getTotalVotingPower (vals *ValidatorSet) int64 {
+	sum := int64(0)
+	for _, val := range vals.Validators {
+		sum += val.VotingPower
+	}
+	return sum
+}
+
+func getTotalProposerPriority (vals *ValidatorSet) int64 {
+	sum := int64(0)
+	for _, val := range vals.Validators {
+		sum += val.ProposerPriority
+	}
+	return sum
+}
+
+func verifyValidatorSet (vals *ValidatorSet) error {
+	// verify that the vals' tvp is set to the sum of the all vals voting powers
+	tvp := getTotalVotingPower(vals)
+	if vals.totalVotingPower != tvp {
+		return fmt.Errorf("expected %d. Got %d, vset=%s", tvp, vals.totalVotingPower, vals)
+	}
+	// verify that validator priorities are centered
+	l := int64(len(vals.Validators))
+	tpp := getTotalProposerPriority(vals)
+	if  tpp > l || tpp < -l {
+		return fmt.Errorf("expected total prio in (-%d, %d). Got %d", l, l, tpp)
+	}
+	// verify that priorities are scaled
+	dist := computeMaxMinPriorityDiff(vals)
+	if dist > K * tvp {
+		return fmt.Errorf("expected prio dist < %d. Got %d", K * tvp, dist)
+	}
+	return nil
+}
+
+func BenchmarkUpdates(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		// Init with n validators
+		n := 100
+		vs := make([]*Validator, n)
+		for j := 0; j < n; j++ {
+			vs[j] = newValidator([]byte(fmt.Sprintf("v%d", j)), 100)
+		}
+		vals := NewValidatorSet(vs)
+		assert.Nil(b, verifyValidatorSet(vals))
+		len := len(vals.Validators)
+		// Add m validators
+		m := 2000
+		nvs := make([]*Validator, m)
+		for j := 0; j < m; j++ {
+			nvs[j] = newValidator([]byte(fmt.Sprintf("v%d", j+len)), 1000)
+		}
```

Comment:
If I understand this correctly, this aims to benchmark the execution of `vals.UpdateWithChangeSet(nvs)`. If so, you could add a line here: 
```suggestion
		}
		b.ResetTimer()
```
As everything until line 813 is initialization for this.

Also, you can remove the assert below (it probably uses reflection and might slow down the benchmark results).

Reply:
My bad. Resetting the timer inside of the loop will make the benchmarks take forever (as they do not "converge").

Reply:
The correct way to do this is probably to move the init code out of the "b.N-loop" (and wrap it with b.StopTimer & b.StartTimer)

Reply:
I made the change, please take a look.